### PR TITLE
[Feature] - Dark mode

### PR DESCRIPTION
**Pull request by paul1893**
Wow ! What an impressive works you did with this library ! I would like to contribute. This is my implementation of the dark mode as asked in the TODO section.

I did some research about UIContextMenu on my jailbroken device that could be interested for you. (see attachements).
It seems Apple used a `UICollectionView` with `UICompositionalLayout` to make his `UIContextMenu`. In addition Apple used `.systemMaterial` for his blur background and has a separator height of `0.33`. I have been able to retreive the exact gray color of the Apple's title context menu (for both light and dark mode).
I added a vibrancy effect on the separator of the table view when it is in dark mode.

```
Changelog:
-Add Dark mode support. It can be either set to automatic ( > iOS 13 - in order to change according to the system preference) or be set to a specific style programatically.
-Change color of "Action" title to the exact UIColor of UIContextMenu title
-Change the popup in/out animation to be a little bit more conform to UIContextMenu
-Change the animation duration
-Update README
-Bump version on podspec
-Remove unnecessary Pods folder from the repository
```

Don't hesitate to challenge me on anything. (I think that my implementation of style for `ContextMenuActionTableViewCell` can be improved instead of passing it the style multiple times but with the `dequeueReusableCell` I can't pass this information to the cell at initialization).
By default if > iOS 13 the default style is `.automatic` and below iOS 13 the style by default is `.light`.

Feel free to merge it .. or not. Have a nice day.

**_PS_**: I didn't make the choice to refactor the stackview title section in a header tableview section even if this change is not a big effort I don't know if this is the wanted behavior (as we know there is a "parallax effect" on the header of a table view with his section's cells).

### CODE REVIEW

**Comment by MarioIannotta** [comment]
Hi @paul1893 and thanks for the PR! I'll definitely have a look :)

File: SwiftyContextMenu/ContextMenu.swift [hunk 1]
```diff
@@ -17,11 +18,18 @@ public struct ContextMenu {
 
     var sourceViewInfo: ContextMenuSourceViewInfo?
 
-    public init(title: String?,
+    public init(style: ContextMenuUserInterfaceStyle? = nil,
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenu.swift]: since this is an optional parameter I would move this after the required parameters `title` and `actions`

File: SwiftyContextMenu/ContextMenu.swift [hunk 2]
```diff
@@ -52,12 +63,17 @@ public enum ContextMenuEvent {
...
+    case light, dark
+}
+
 public struct ContextMenuAnimation {
     let sourceViewBounceRange: ClosedRange<CGFloat>
     let optionsViewBounceRange: ClosedRange<CGFloat>
         
     public init(sourceViewBounceRange: ClosedRange<CGFloat> = 0.9...1.15,
-                optionsViewBounceRange: ClosedRange<CGFloat> = 0.9...1) {
+                optionsViewBounceRange: ClosedRange<CGFloat> = 0.05...1) {
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenu.swift:76]: is this a debug-leftover or is this change required to have a more "iOS-like animation"?

File: SwiftyContextMenu/ContextMenuActionTableViewCell.swift [hunk 3]
```diff
@@ -8,25 +8,86 @@
...
+    
     private var rightImageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 25, height: 25))
-
+    private var lightSelectedBackgroundView: UIVisualEffectView!
+    private var darkSelectedBackgroundView: UIView!
+    private var separatorView: ContextMenuSeparatorView!
+    private var style: ContextMenuUserInterfaceStyle = .light {
+        didSet {
+            updateSelectedBackgroundView()
+            separatorView.style = style
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuActionTableViewCell.swift]: I don't feel very safe with this code, all properties are private so we are in control of their life-cycle but I would use an optional here, just in case :) 
`separatorView?.style = style`

File: SwiftyContextMenu/ContextMenuActionTableViewCell.swift [hunk 4]
```diff
@@ -8,25 +8,86 @@
...
+        case .light:
+            textLabel?.textColor = action.tintColor
+            rightImageView.tintColor = action.tintColor
+        case .dark:
+            textLabel?.textColor = action.tintColorDark
+            rightImageView.tintColor = action.tintColorDark
+        }
+    }
+    
+    fileprivate func prepareSelectedBackgroundView() {
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuActionTableViewCell.swift]: this is more an "add" than a "prepare", right? plus I'm not sure why this is `fileprivate` instead of `private`.
I would change this into: `private func addSelectedBackgroundView()`

File: SwiftyContextMenu/ContextMenuActionTableViewCell.swift [hunk 5]
```diff
@@ -8,25 +8,86 @@
...
+    
     override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
         super.init(style: style, reuseIdentifier: reuseIdentifier)
         backgroundColor = .clear
         rightImageView.contentMode = .scaleAspectFit
         accessoryView = rightImageView
+        
+        prepareSelectedBackgroundView()
+        
+        separatorView = ContextMenuSeparatorView(frame: CGRect(x: 0, y: 0, width: bounds.width, height: 0.33), style: self.style)
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuActionTableViewCell.swift]: can we update this code to use autolayout instead of frame layout? 
In this way we don't need to do any size updates in `layoutSubviews` ourselfs.

```
let separatorView = ContextMenuSeparatorView(frame: .zero)
separatorView. translatesAutoresizingMaskIntoConstraints = false
NSLayoutConstraint.activate([
separatorView.leading.....
])
```

File: SwiftyContextMenu/ContextMenuBackgroundBlurView.swift [hunk 6]
```diff
@@ -0,0 +1,44 @@
+//
+//  ContextMenuBackgroundBlurView.swift
+//  SwiftyContextMenu
+//
+//  Created by Paul Bancarel on 28/11/2020.
+//
+
+final class ContextMenuBackgroundBlurView : IntensityVisualEffectView {
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuBackgroundBlurView.swift]: the coma should be near the type name
`class ContextMenuBackgroundBlurView: IntensityVisualEffectView`

File: SwiftyContextMenu/ContextMenuBackgroundBlurView.swift [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        effect = ContextMenuBackgroundBlurView.blurEffect(style)
+    }
+    
+    fileprivate static func blurEffect(_ style: ContextMenuUserInterfaceStyle) -> UIBlurEffect {
+        switch style {
+        case .automatic:
+            if #available(iOS 13.0, *) {
+                return UIBlurEffect(style: .light)
+            } else {
+                fatalError("Cannot have an automatic blur effect below iOS 13.")
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuBackgroundBlurView.swift]: wait, instead of making the app crash, can't we print and return some default value?

File: SwiftyContextMenu/ContextMenuViewController.swift [hunk 8]
```diff
@@ -14,24 +14,25 @@ protocol ContextMenuViewControllerDelegate: class {
...
-    private var blurView: UIVisualEffectView = UIVisualEffectView(effect: UIBlurEffect(style: .light))
-    private var overlayView = UIView(frame: .zero)
-    private var snapshotImageView = UIImageView(frame: .zero)
-    private var contextMenuTableView = ContextMenuTableView()
-    private var contextMenuView = UIView()
+    private let blurView: UIVisualEffectView
+    private let overlayView = UIView(frame: .zero)
+    private let snapshotImageView = UIImageView(frame: .zero)
+    private let contextMenuTableView = ContextMenuTableView()
+    private let contextMenuView = UIView()
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuViewController.swift:24]: really like these changes! thank you 👍

File: SwiftyContextMenu/ContextMenuViewController.swift [hunk 9]
```diff
@@ -14,24 +14,25 @@ protocol ContextMenuViewControllerDelegate: class {
...
-    private var contextMenu: ContextMenu!
+    private let contextMenu: ContextMenu
     private weak var delegate: ContextMenuViewControllerDelegate?
 
-    private var blurView: UIVisualEffectView = UIVisualEffectView(effect: UIBlurEffect(style: .light))
-    private var overlayView = UIView(frame: .zero)
-    private var snapshotImageView = UIImageView(frame: .zero)
-    private var contextMenuTableView = ContextMenuTableView()
-    private var contextMenuView = UIView()
+    private let blurView: UIVisualEffectView
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuViewController.swift]: I think the type of this property should be `ContextMenuBackgroundBlurView`

File: SwiftyContextMenu/ContextMenuBlurView.swift [hunk 10]
```diff
@@ -0,0 +1,48 @@
+//
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuBlurView.swift:1]: this class shares pretty much all the code with `ContextMenuBackgroundBlurView`. 

Maybe we can create a base class `ContextMenuBlurView` and have two child classes `ContextMenuBackgroundBlurView: ContextMenuBlurView `, `ContextMenuContentBlurView: ContextMenuBlurView` 🤔

File: SwiftyContextMenu/ContextMenuTitleLabel.swift [hunk 11]
```diff
@@ -0,0 +1,42 @@
+//
+//  ContextMenuTitleLabel.swift
+//  SwiftyContextMenu
+//
+//  Created by Paul Bancarel on 28/11/2020.
+//
+
+final class ContextMenuTitleLabel : UILabel {
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuTitleLabel.swift]: the coma should go near the type
`class ContextMenuTitleLabel: UILabel`

**Review by MarioIannotta — CHANGES_REQUESTED** [review]
This looks very good, I need to have a better look to the animation changes but thank you so far!

I really appreciate it!

**Comment by MarioIannotta** [comment]
There's one minor but very useful feature that's broken in this branch. If a source view is partially hidden (for example it is in a scrollview) the animation was written in such a way that the source view will move to be completely visible during the animation (similar to what the Apple component already does).

You can reproduce that behavior on the main branch changing the horizontal constraints of the main stack view to something like (-40, -40)

## Main branch
 


## Dark mode branch

File: SwiftyContextMenu/ContextMenu.swift (hunk 2 above)
Comment (paul1893) [reply SwiftyContextMenu/ContextMenu.swift:76]: Yes this is a change to have a more "iOS-like animation". My simulator is on iOS 14. I don't know if it's an Apple's animation changes from 13 to 14 but the animation on iOS 14 is more like scaling and translate from the view that has been touched. Don't have a device on iOS 13 with me to know if it was the case on iOS 13 too 🤷‍♂️.

File: SwiftyContextMenu/ContextMenuActionTableViewCell.swift (hunk 3 above)
Comment (paul1893) [reply SwiftyContextMenu/ContextMenuActionTableViewCell.swift]: You're right we don't how it can evolve in the future. Safety first.

File: SwiftyContextMenu/ContextMenuBackgroundBlurView.swift (hunk 6 above)
Comment (paul1893) [reply SwiftyContextMenu/ContextMenuBackgroundBlurView.swift]: 👍 I'm really like that attention to detail

File: SwiftyContextMenu/ContextMenuBackgroundBlurView.swift (hunk 7 above)
Comment (paul1893) [reply SwiftyContextMenu/ContextMenuBackgroundBlurView.swift]: I made the change but normally it's impossible to pass throught that code because `.automatic` is iOS13 + due to:
```
public enum ContextMenuUserInterfaceStyle {
    @available(iOS 13, *) case automatic
    case light, dark
}
```
But at compile time swift doesn't know that. If i'm in a `switch case .automatic` then I'm necessarily on iOS 13+. Swift asks me to check but it's useless.
Do you have an idea to change the implementation in order to make the compiler understand that ?

File: SwiftyContextMenu/ContextMenuViewController.swift (hunk 9 above)
Comment (paul1893) [reply SwiftyContextMenu/ContextMenuViewController.swift]: Well done 👍

File: SwiftyContextMenu/ContextMenuBlurView.swift (hunk 10 above)
Comment (paul1893) [reply SwiftyContextMenu/ContextMenuBlurView.swift:1]: Yes this makes sense

**Comment by paul1893** [comment]
Thanks @MarioIannotta for this rich code review! I really appreciate that you took that time 👍

Regarding the bug about the source view that moves to be completely visible during the animation, I restored the computation you made to handle this case. This fix the issue. And I added a condition that made the computation happens only if the source view is out of the screen when the animation starts.
Do not hesitate to tell me if this is OK for you elsewhere I will remove this condition.

Furthermore, really do not hesitate if you see something else to correct.

File: SwiftyContextMenu/ContextMenuBackgroundBlurView.swift [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    
+    override class func blurEffect(_ style: ContextMenuUserInterfaceStyle) -> UIVisualEffect {
+        switch style {
+        case .automatic:
+            if #available(iOS 13.0, *) {
+                return UIBlurEffect(style: .light)
+            } else {
+                print("Cannot have an automatic blur effect below iOS 13.")
+                return UIBlurEffect(style: .light)
+            }
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuBackgroundBlurView.swift]: Well at this point this could just be 
```
return UIBlurEffect(style: .light)
```
right?

File: SwiftyContextMenu/ContextMenuBlurView.swift [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+//  ContextMenuBlurView.swift
+//  SwiftyContextMenu
+//
+//  Created by Paul Bancarel on 28/11/2020.
+//
+
+class ContextMenuBlurView: IntensityVisualEffectView {
+    
+    internal class var intensity: CGFloat { 1.0 }
```
Comment (MarioIannotta) [inline SwiftyContextMenu/ContextMenuBlurView.swift]: I don't think internal is required since it is the default access control qualifier

**Review by MarioIannotta — COMMENTED** [review]
The code looks good besides some minor picky comments, if you have time please fix them otherwise I will take care of the polishing when releasing a new pod.

I'll test the UX and some edge-cases a little bit and if everything is fine we can merge this PR.

Thank you so much for you effort!

**Comment by paul1893** [comment]
> The code looks good besides some minor picky comments, if you have time please fix them otherwise I will take care of the polishing when releasing a new pod.
> 
> I'll test the UX and some edge-cases a little bit and if everything is fine we can merge this PR.
> 
> Thank you so much for you effort!

Thanks. I will fix them tomorrow morning when I will get to work ;)

**Comment by MarioIannotta** [comment]
The UX overall looks very good, I found one little issue though. 
It looks like the vertical padding of the options is wrong.
Here's an example, check the space between the separators and the "Looooooooong title" label.

| New |  Old |
| ----- | ----|
|  |  |

**Comment by paul1893** [comment]
Just pushed the last changes requested :)

Moreover this changes reminds me that table view cells automatically resize to fit their content by default only from iOS 11 right ?
As SwiftyContextMenu supports iOS 10+. Should we add (for iOS10) ?
```
contextMenuTableView.rowHeight = UITableView.automaticDimension
contextMenuTableView.estimatedRowHeight = 44
```
I do not have any device on iOS 10 with me and [XCode 12 does not support iOS 10 simulator on macOS BigSur](https://stackoverflow.com/a/65042303) :(
I you think I'm wrong or if you can test it on iOS10, I will delete this two unnecessary lines.

File: SwiftyContextMenu/ContextMenuActionTableViewCell.swift [hunk 14]
```diff
@@ -68,14 +68,17 @@ class ContextMenuActionTableViewCell: UITableViewCell {
...
-        addSubview(separatorView)
         separatorView.translatesAutoresizingMaskIntoConstraints = false
-        NSLayoutConstraint.activate([
-            separatorView.heightAnchor.constraint(equalToConstant: 0.33),
-            separatorView.topAnchor.constraint(equalTo: topAnchor),
-            separatorView.leadingAnchor.constraint(equalTo: leadingAnchor),
-            separatorView.trailingAnchor.constraint(equalTo: trailingAnchor)
-        ])
+        contentView.addSubview(separatorView)
+        DispatchQueue.main.async {
```
Comment (paul1893) [inline SwiftyContextMenu/ContextMenuActionTableViewCell.swift:73]: Sync with the next frame otherwise it's too soon to set constraints for the cell and this warning's pop in logs.

```
[Warning] Warning once only: Detected a case where constraints ambiguously suggest a height of zero for a table view cell's content view. We're considering the collapse unintentional and using standard height instead. Cell: <SwiftyContextMenu.ContextMenuActionTableViewCell: 0x7fd698c1e7b0; baseClass = UITableViewCell; frame = (0 0; 250 44); text = 'Looooooooooooong title'; autoresize = W; layer = <CALayer: 0x600002ca11e0>>
```
As other solutions we could also: 
- Move the `NSLayoutConstraint.activate` in `layoutSubviews()` (but I think it will be less efficient).
- Abandon separatorView's auto-layout for frame-layout
- Other solution that you may think of ?

Comment (MarioIannotta) [reply SwiftyContextMenu/ContextMenuActionTableViewCell.swift:73]: I like this solution 👍

**Comment by MarioIannotta** [comment]
I tested the UX and it looks very good! I'll merge the PR 🎉
Thank you very much!

**Review by MarioIannotta — APPROVED** [review]

**Comment by MarioIannotta** [comment]
This PR has been shipped with the version 0.1.5 - tag [0.1.5](https://github.com/MarioIannotta/SwiftyContextMenu/releases/tag/0.1.5).
```
--------------------------------------------------------------------------------
 🎉  Congrats

 🚀  SwiftyContextMenu (0.1.5) successfully published
 📅  December 1st, 19:21
 🌎  https://cocoapods.org/pods/SwiftyContextMenu
 👍  Tell your friends!
--------------------------------------------------------------------------------
```
